### PR TITLE
Bitcoin

### PR DESCRIPTION
**Pull request by rmui1**

### CODE REVIEW

File: src/org/usfirst/frc/team694/robot/OI.java [hunk 1]
```diff
@@ -21,6 +21,7 @@
     public OI() {
          driverGamepad = new Gamepad(RobotMap.DRIVER_GAMEPAD_PORT, GamepadSwitchMode.PS4);
          operatorGamepad = new Gamepad(RobotMap.OPERATOR_GAMEPAD_PORT, GamepadSwitchMode.SWITCH_X);
+         operatorGamepad.getRightButton().whenPressed()
```
Comment (TacoTechnica) [inline src/org/usfirst/frc/team694/robot/OI.java]: This won't compile. Either delete this or finish this line

File: src/org/usfirst/frc/team694/robot/commands/BITCOINCheckCommand.java [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+import org.usfirst.frc.team694.robot.Robot;
+import org.usfirst.frc.team694.util.IRSensor;
+
+import edu.wpi.first.wpilibj.command.Command;
+import edu.wpi.first.wpilibj.command.InstantCommand;
+
+/**
+ *
+ */
+public class BITCOINCheckCommand extends InstantCommand {
```
Comment (TacoTechnica) [inline src/org/usfirst/frc/team694/robot/commands/BITCOINCheckCommand.java]: This should not be an InstantCommand.

This command should never stop and have an execute method that checks our IR sensor and a button to determine whether to run the Bitcoin command.

File: src/org/usfirst/frc/team694/robot/commands/BITCOINCheckCommand.java [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+    public Command BITCOINCommand;
+    public BITCOINCheckCommand() {
+        super();
+        this.BITCOINCommand = new BITCOINCommand();
+        // Use requires() here to declare subsystem dependencies
+        // eg. requires(chassis);
+    }
+
+    // Called just before this Command runs the first time
+    protected void initialize() {
```
Comment (TacoTechnica) [inline src/org/usfirst/frc/team694/robot/commands/BITCOINCheckCommand.java:20]: The code in `initialize()` should be moved to `execute()`

File: src/org/usfirst/frc/team694/robot/commands/BITCOINCommand.java [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+        addSequential(new AcquirerTightenHoldCommand());
+        addSequential(new AcquirerFlipUpCommand());
+        addSequential(new LiftMaxDownCommand());
+        //TODO when merge to branch with command, uncomment
+        //addSequential(new LiftMoveUpHeightCommand(height));
+        addSequential(new AcquirerLoosenHoldCommand());
+        addSequential(new GrabberCloseCommand());
+        }
+    
+    public void initialize() {
```
Comment (TacoTechnica) [inline src/org/usfirst/frc/team694/robot/commands/BITCOINCommand.java:24]: You shouldn't have `initialize()` and `execute()` in CommandGroups. See my comments from the `BITCOINCheckCommand`

File: src/org/usfirst/frc/team694/robot/commands/auton/DrivetrainCrashCommand.java [hunk 5]
```diff
@@ -0,0 +1,39 @@
+package org.usfirst.frc.team694.robot.commands.auton;
+
+import org.usfirst.frc.team694.robot.Robot;
+
+import edu.wpi.first.wpilibj.command.Command;
+
+/**
+ *
+ */
+public class DrivetrainCrashCommand extends Command {
```
Comment (TacoTechnica) [inline src/org/usfirst/frc/team694/robot/commands/auton/DrivetrainCrashCommand.java]: Hmmm... Why is this here?

File: src/org/usfirst/frc/team694/robot/commands/auton/DrivetrainMoveToLineCommand.java [hunk 6]
```diff
@@ -10,10 +10,10 @@
 public class DrivetrainMoveToLineCommand extends Command {
     private boolean isReached = false; 
     private double distance = 0;
-    private double speedMode = 0;
+    private double speedNumber = 0;// 0,1, or 2, repesents speeds 0.25,0.5,0.75 respectively.
```
Comment (TacoTechnica) [inline src/org/usfirst/frc/team694/robot/commands/auton/DrivetrainMoveToLineCommand.java]: Pls update, no speedNumber or speedMode. Just speed (0 to 1)

File: src/org/usfirst/frc/team694/robot/commands/auton/DrivetrainMoveToLineCommand.java [hunk 7]
```diff
@@ -27,13 +27,13 @@ protected void initialize() {
...
-        Robot.drivetrain.tankDrive((speedMode + 1) * 0.25, (speedMode + 1) * 0.25);
-        isReached = Robot.drivetrain.isOnLine((int) speedMode);
+        Robot.drivetrain.tankDrive((speedNumber + 1) * 0.25, (speedNumber + 1) * 0.25);
+        isReached = Robot.drivetrain.isOnLine((int) speedNumber);
     }
 
     // Make this return true when this Command no longer needs to run execute()
     protected boolean isFinished() {
-        return isReached || (Math.max(Robot.drivetrain.getRightEncoderDistance(),Robot.drivetrain.getLeftEncoderDistance()) > (distance + 5));
+        return isReached || (Math.max(Robot.drivetrain.getRightEncoderDistance(),Robot.drivetrain.getLeftEncoderDistance()) > (distance));
```
Comment (TacoTechnica) [inline src/org/usfirst/frc/team694/robot/commands/auton/DrivetrainMoveToLineCommand.java]: Pls fix (use `Robot.drivetrain.getEncoderDistance()`)

File: src/org/usfirst/frc/team694/robot/OI.java (hunk 1 above)
Comment (PrathamRawat) [inline src/org/usfirst/frc/team694/robot/OI.java]: Finish this, this line will cause an error. Either delete it, or make it do something, I still don't get the point of this.

File: src/org/usfirst/frc/team694/robot/Robot.java [hunk 8]
```diff
@@ -53,6 +55,7 @@ public void robotInit() {
         grabber = new Grabber();
         lift = new Lift();
         oi = new OI();
+        isBITCOINAutomation = true;
```
Comment (PrathamRawat) [inline src/org/usfirst/frc/team694/robot/Robot.java]: The robotInit() line is not being read, it's included in the @ above it. This will cause an error.

File: src/org/usfirst/frc/team694/robot/commands/BITCOINCheckCommand.java [hunk 9]
```diff
@@ -0,0 +1,29 @@
...
+        if (Robot.isBITCOINAutomation) {
+            if (IRSensor.isCubeDetected()) {
```
Comment (PrathamRawat) [inline src/org/usfirst/frc/team694/robot/commands/BITCOINCheckCommand.java]: Should this be using the IR sensor class directly, or should it be going through the Acquirer subsystem, or even Robot. In Rafael, we use the IR sensor port, but we only have one sensor. I think it would be a good idea to do this in a more Object Oriented way, with an IR sensor object in the Acquirer class, and methods in the Acquirer class to get that information, not accessing the IR sensor class directly. I would suggest you discuss this more with Adris.

File: src/org/usfirst/frc/team694/robot/commands/ToggleBITCOINAutomationCommand.java [hunk 10]
```diff
@@ -0,0 +1,22 @@
+package org.usfirst.frc.team694.robot.commands;
+
+import org.usfirst.frc.team694.robot.Robot;
+
+import edu.wpi.first.wpilibj.command.InstantCommand;
+
+/**
+ *
+ */
+public class ToggleBITCOINAutomationCommand extends InstantCommand {
```
Comment (PrathamRawat) [inline src/org/usfirst/frc/team694/robot/commands/ToggleBITCOINAutomationCommand.java]: This is all cool and good, but based off of the Operator Pad layout, there will not be a toggle for turning BITCOIN on and off, but rather two individual buttons, one for turning it off, and one for turning it on. You don't have to delete this command, but make two more to reflect that layout.

File: src/org/usfirst/frc/team694/robot/subsystems/Acquirer.java [hunk 11]
```diff
@@ -15,22 +17,23 @@
...
+    public Solenoid acquirerSqueezeSolenoid; 
     public SpeedControllerGroup acquirerMotors;
 
     // Put methods for controlling this subsystem
     // here. Call these from Commands.
 
     public void initDefaultCommand() {
         // Set the default command for a subsystem here.
-        //setDefaultCommand(new MySpecialCommand());
+        setDefaultCommand(new BITCOINCheckCommand());
```
Comment (PrathamRawat) [inline src/org/usfirst/frc/team694/robot/subsystems/Acquirer.java:33]: Again, think about IR sensors in the Acquirer class.

**Review by PrathamRawat — CHANGES_REQUESTED** [review]
Read the specific comments for more information. This is mostly cool and good, but there are a few small errors in need of fixing, and a little more discussion in how we implement this perhaps.

**Review by TacoTechnica — CHANGES_REQUESTED** [review]
Change your BITCOINCheckCommand to be a regular command that constantly checks for the IR Sensor in the execute method.

Change your BITCOINCommand to have no initialize() and execute() method (you can move these to BITCOINCheckCommand.java)

Also check over Pratham's comments. If you have any questions, don't hesitate to ask me during a meeting!

**Review by huiminwu — APPROVED** [review]
Good to go!

**Review by TacoTechnica — APPROVED** [review]
Good to Go!

**Review by PrathamRawat — APPROVED** [review]
